chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,263 @@
+package mir
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/multiformats/go-multiaddr"
+	"go4.org/sort"
+
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+type Messages []byte
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships    []map[t.NodeID]t.NodeAddress
+	membershipLock sync.Mutex
+
+	// Channel to send messages to Eudico.
+	ChainNotify chan []Messages
+
+	MirManager *Manager
+
+	reconfigurationVotes map[string]uint64
+}
+
+func NewStateManager(initialMembership map[t.NodeID]t.NodeAddress, m *Manager) *StateManager {
+	// Initialize the membership for the first epochs.
+	// We use configOffset+2 memberships to account for:
+	// - The first epoch (epoch 0)
+	// - The configOffset epochs that already have a fixed membership (epochs 1 to configOffset)
+	// - The membership of the following epoch (configOffset+1) initialized with the same membership,
```

Comment:
It seems to me that we don't need the extra membership here. It was only used in the old version of the chat app, because the new membership was being progressively constructed (and even there it was actually an ugly solution to keep it directly in the slice).
Would it make sense to have the memberships contain only the agreed-upon memberships in the case of the state manager? Whenever a new membership is agreed upon, it could be just appended here.

Reply:
I implemented removing memberships for the old epochs. Did it fix this comment?

Reply:
What I meant was slightly different, but it's ok the way it is now.